File: src/ingestion/normalize.ts
import type { Category } from '@/types/article'
import type { Source } from './sources'

export interface NormalizedArticle {
  title: string
  body_excerpt: string | null
  source_name: string
  source_url: string
  author: string | null
  published_at: string // ISO 8601
  raw_category: string
  significance_base: number
}

// Keyword map for category classification. Title match beats body match.
const CATEGORY_KEYWORDS: Record<Exclude<Category, 'uncategorized'>, string[]> = {
  'model-releases': [
    'release', 'launches', 'launched', 'introducing', 'new model', 'gpt-', 'claude',
    'gemini', 'llama', 'mistral', 'grok', 'phi-', 'qwen', 'deepseek', 'weights',
    'checkpoint', 'fine-tun', 'instruct', 'chat model',
  ],
  'research-papers': [
    'arxiv', 'paper', 'research', 'study', 'we propose', 'we present', 'benchmark',
    'dataset', 'training', 'preprint', 'journal', 'conference', 'icml', 'neurips',
    'iclr', 'acl', 'emnlp',
  ],
  'company-news': [
    'funding', 'raises', 'acquisition', 'acquires', 'partnership', 'partnership',
    'valuation', 'billion', 'million', 'ceo', 'executive', 'leadership', 'series',
    'venture', 'ipo', 'layoffs', 'hired',
  ],
  'product-launches': [
    'api', 'integration', 'feature', 'update', 'plugin', 'app', 'tool', 'platform',
    'service', 'product', 'available now', 'generally available', 'ga ', 'beta',
    'preview', 'copilot', 'assistant', 'chatbot',
  ],
  'regulation-policy': [
    'regulation', 'policy', 'law', 'congress', 'senate', 'eu ', 'government',
    'act ', 'ban', 'safety', 'alignment', 'risk', 'guidelines', 'compliance',
    'gdpr', 'executive order', 'legislation',
  ],
  'hardware-infrastructure': [
    'gpu', 'tpu', 'chip', 'nvidia', 'amd', 'intel', 'hardware', 'compute',
    'data center', 'inference', 'h100', 'h200', 'blackwell', 'cuda', 'cluster',
    'supercomputer', 'training infrastructure',
  ],
  'open-source': [
    'open source', 'open-source', 'github', 'release v', 'v0.', 'v1.', 'v2.',
    'repo', 'repository', 'hugging face', 'weights', 'community', 'apache',
    'mit license', 'ollama', 'llama.cpp', 'vllm', 'langchain',
  ],
  'opinion-analysis': [
    'opinion', 'analysis', 'commentary', 'editorial', 'perspective', 'why ',
    'how ', 'what ', 'should', 'future of', 'impact of', 'the case for',
    'the case against', 'reflections',
  ],
}

// Titles matching these patterns are financial content, not AI news — drop them
const FINANCIAL_NOISE_PATTERNS = [
  'stock', 'stocks', 'shares', 'buy hand over fist', 'investor', 'portfolio',
  'etf', 'nasdaq', 'wall street', 'earnings', 'valuation', 'market cap',
  'price target', 'analyst', 'bull case', 'bear case', 'to buy in',
  'top pick', 'strong buy', 'undervalued', 'overvalued',
  // Additional investment/financial noise patterns
  'sell-off', 'selloff', 'worth a fortune', 'getting cheaper',
  'could be worth', 'motley fool', 'the motley fool',
  'trillion in sales', 'revenue forecast', 'market share',
  'buy before', 'before it', "don't miss", 'missed out',
  'billionaire', 'hedge fund',
]

/** Returns true if the title looks like financial noise rather than AI news. */
export function isFinancialNoise(title: string): boolean {
  const t = title.toLowerCase()
  return FINANCIAL_NOISE_PATTERNS.some((p) => t.includes(p))
}

/**
 * Returns true if a GitHub release looks like a rolling/incremental build rather
 * than a named release worth showing (e.g. llama.cpp "b8560 — b8560").
 * Semver tags (containing a dot) are always kept.
 */
export function isGitHubRollingBuild(tagName: string, releaseName: string | null): boolean {
  // Semver tags (v4.50.0, 0.8.1) are always meaningful — keep them
  if (tagName.includes('.')) return false

  // Pure hex hash — a raw commit SHA used as a tag (e.g. "abc1234f")
  if (/^[a-f0-9]{5,}$/i.test(tagName)) return true

  // llama.cpp rolling build numbers: "b8560", "b9001"
  if (/^b\d+$/.test(tagName)) return true

  // Tag and name are identical with no version structure — no extra context
  if (releaseName !== null && releaseName === tagName) return true

  return false
}

export function categorize(title: string, body: string | null): Category {
  const titleLower = title.toLowerCase()
  const bodyLower = (body ?? '').toLowerCase()

  for (const [category, keywords] of Object.entries(CATEGORY_KEYWORDS)) {
    if (keywords.some((kw) => titleLower.includes(kw))) {
      return category as Category
    }
  }
  for (const [category, keywords] of Object.entries(CATEGORY_KEYWORDS)) {
    if (keywords.some((kw) => bodyLower.includes(kw))) {
      return category as Category
    }
  }
  return 'uncategorized'
}

/** Strip HTML tags from a string */
function stripHtml(html: string): string {
  return html.replace(/<[^>]*>/g, ' ').replace(/\s+/g, ' ').trim()
}

/** Truncate to at most 150 words (copyright compliance) */
function truncateWords(text: string, maxWords = 150): string {
  const words = text.trim().split(/\s+/)
  if (words.length <= maxWords) return text
  return words.slice(0, maxWords).join(' ') + '…'
}

export function normalizeExcerpt(raw: string | null | undefined): string | null {
  if (!raw) return null
  const stripped = stripHtml(raw)
  if (!stripped) return null
  return truncateWords(stripped)
}

/** Base significance score. Social signals added in Phase 2. */
export function scoreSignificance(authority: number): number {
  return authority
}

function coerceAuthor(val: unknown): string | null {
  if (!val) return null
  if (typeof val === 'string') return val.trim() || null
  if (typeof val === 'object' && val !== null) {
    const obj = val as Record<string, unknown>
    return coerceAuthor(obj.name ?? obj.text ?? obj['#text'])
  }
  return null
}

export function buildNormalizedArticle(
  params: {
    title: string
    url: string
    excerpt: string | null | undefined
    author: unknown
    publishedAt: string | Date | number | null | undefined
  },
  source: Source,
): NormalizedArticle | null {
  const title = params.title?.trim()
  if (!title || !params.url) return null

  const body_excerpt = normalizeExcerpt(params.excerpt)
  const published_at = toISOString(params.publishedAt)
  if (!published_at) return null

  const raw_category = categorize(title, body_excerpt)

  return {
    title,
    body_excerpt,
    source_name: source.name,
    source_url: params.url,
    author: coerceAuthor(params.author),
    published_at,
    raw_category,
    significance_base: scoreSignificance(source.authority),
  }
}

function toISOString(val: string | Date | number | null | undefined): string | null {
  if (!val) return null
  try {
    const d = new Date(typeof val === 'number' ? val * 1000 : val)
    if (isNaN(d.getTime())) return null
    return d.toISOString()
  } catch {
    return null
  }
}

File: pipeline/src/ainews/processing/normalize.py
import re
from datetime import datetime, timezone
from html.parser import HTMLParser

from ..models import NormalizedArticle
from ..sources import Source


CATEGORY_KEYWORDS: dict[str, list[str]] = {
    "model-releases": [
        "release", "launches", "launched", "introducing", "new model", "gpt-", "claude",
        "gemini", "llama", "mistral", "grok", "phi-", "qwen", "deepseek", "weights",
        "checkpoint", "fine-tun", "instruct", "chat model",
    ],
    "research-papers": [
        "arxiv", "paper", "research", "study", "we propose", "we present", "benchmark",
        "dataset", "training", "preprint", "journal", "conference", "icml", "neurips",
        "iclr", "acl", "emnlp",
    ],
    "company-news": [
        "funding", "raises", "acquisition", "acquires", "partnership",
        "valuation", "billion", "million", "ceo", "executive", "leadership", "series",
        "venture", "ipo", "layoffs", "hired",
    ],
    "product-launches": [
        "api", "integration", "feature", "update", "plugin", "app", "tool", "platform",
        "service", "product", "available now", "generally available", "ga ", "beta",
        "preview", "copilot", "assistant", "chatbot",
    ],
    "regulation-policy": [
        "regulation", "policy", "law", "congress", "senate", "eu ", "government",
        "act ", "ban", "safety", "alignment", "risk", "guidelines", "compliance",
        "gdpr", "executive order", "legislation",
    ],
    "hardware-infrastructure": [
        "gpu", "tpu", "chip", "nvidia", "amd", "intel", "hardware", "compute",
        "data center", "inference", "h100", "h200", "blackwell", "cuda", "cluster",
        "supercomputer", "training infrastructure",
    ],
    "open-source": [
        "open source", "open-source", "github", "release v", "v0.", "v1.", "v2.",
        "repo", "repository", "hugging face", "weights", "community", "apache",
        "mit license", "ollama", "llama.cpp", "vllm", "langchain",
    ],
    "opinion-analysis": [
        "opinion", "analysis", "commentary", "editorial", "perspective", "why ",
        "how ", "what ", "should", "future of", "impact of", "the case for",
        "the case against", "reflections",
    ],
}

AI_RELEVANCE_KEYWORDS = [
    # Multi-word or unambiguous single-word AI terms
    "artificial intelligence", "machine learning", "deep learning",
    "language model", "llm", "neural network", "neural net",
    "generative model", "foundation model",
    # Specific model/product names
    "chatgpt", "gpt-", "claude", "gemini", "llama", "mistral",
    "grok", "phi-", "qwen", "deepseek", "copilot", "dall-e",
    "midjourney", "stable diffusion", "sora", "glm-", "codex",
    # AI company names (unambiguous even without "AI" suffix)
    "openai", "anthropic", "deepmind", "hugging face", "huggingface",
    "sakana", "sarvam",
    # Technical terms
    "fine-tun", "rlhf", "diffusion model", "embedding model", "chatbot", "deepfake",
]

# Word-boundary regex for "AI" as a standalone word (avoids "email", "paid", "Taiwan")
_AI_WORD_RE = re.compile(r'\bai\b', re.IGNORECASE)


def is_ai_relevant(title: str, body: str | None) -> bool:
    """Return True if an article is actually about AI/ML topics.

    Used for sources with ai_filter=True to block non-AI articles that slip
    through categorization (e.g. titles with 'billion'/'million' get tagged
    'company-news' even when they're about GDP or VR hardware).

    Strategy: word-boundary regex for "AI" as a standalone word handles the
    majority of cases cleanly (avoids false matches on "email", "paid", etc.).
    Falls back to specific keyword list for AI articles that don't use the word
    "AI" but reference model names, company names, or technical terms.
    """
    combined = (title + " " + (body or "")).lower()
    if _AI_WORD_RE.search(combined):
        return True
    return any(kw in combined for kw in AI_RELEVANCE_KEYWORDS)


FINANCIAL_NOISE_PATTERNS = [
    "stock", "stocks", "shares", "buy hand over fist", "investor", "portfolio",
    "etf", "nasdaq", "wall street", "earnings", "valuation", "market cap",
    "price target", "analyst", "bull case", "bear case", "to buy in",
    "top pick", "strong buy", "undervalued", "overvalued",
    "sell-off", "selloff", "worth a fortune", "getting cheaper",
    "could be worth", "motley fool", "the motley fool",
    "trillion in sales", "revenue forecast", "market share",
    "buy before", "before it", "don't miss", "missed out",
    "billionaire", "hedge fund",
]


def is_english(text: str) -> bool:
    """Return True if text is likely English.

    langdetect is unreliable on short text (< ~15 words) and mis-fires often on
    technical AI titles full of proper nouns. Strategy:
    - Short text → assume English (safe default).
    - Longer text → run langdetect; reject only if confidently non-English (score > 0.98).
    This keeps false-positive deletions near zero while still catching long
    Spanish/French/German articles like Bloomberg's international content.
    """
    words = text.split()
    if len(words) < 10:
        return True  # Too short to classify reliably — keep it
    try:
        from langdetect import detect_langs
        results = detect_langs(text)
        # Only reject if the top language is non-English with very high confidence
        if results and results[0].lang != "en" and results[0].prob > 0.98:
            return False
        return True
    except Exception:
        return True  # Default to include on any error


def is_financial_noise(title: str) -> bool:
    t = title.lower()
    return any(p in t for p in FINANCIAL_NOISE_PATTERNS)


def is_github_rolling_build(tag_name: str, release_name: str | None) -> bool:
    if "." in tag_name:
        return False
    if re.match(r"^[a-f0-9]{5,}$", tag_name, re.IGNORECASE):
        return True
    if re.match(r"^b\d+$", tag_name):
        return True
    if release_name is not None and release_name == tag_name:
        return True
    return False


def categorize(title: str, body: str | None) -> str:
    title_lower = title.lower()
    body_lower = (body or "").lower()
    for category, keywords in CATEGORY_KEYWORDS.items():
        if any(kw in title_lower for kw in keywords):
            return category
    for category, keywords in CATEGORY_KEYWORDS.items():
        if any(kw in body_lower for kw in keywords):
            return category
    return "uncategorized"


class _HTMLStripper(HTMLParser):
    def __init__(self) -> None:
        super().__init__()
        self._parts: list[str] = []

    def handle_data(self, data: str) -> None:
        self._parts.append(data)

    def get_text(self) -> str:
        return " ".join(self._parts)


def strip_html(html: str) -> str:
    stripper = _HTMLStripper()
    stripper.feed(html)
    text = stripper.get_text()
    return re.sub(r"\s+", " ", text).strip()


def truncate_words(text: str, max_words: int = 150) -> str:
    words = text.strip().split()
    if len(words) <= max_words:
        return text
    return " ".join(words[:max_words]) + "…"


def normalize_excerpt(raw: str | None) -> str | None:
    if not raw:
        return None
    stripped = strip_html(raw)
    if not stripped:
        return None
    return truncate_words(stripped)


def to_iso_string(val: str | int | float | datetime | None) -> str | None:
    if val is None:
        return None
    try:
        if isinstance(val, (int, float)):
            dt = datetime.fromtimestamp(val, tz=timezone.utc)
        elif isinstance(val, datetime):
            dt = val if val.tzinfo else val.replace(tzinfo=timezone.utc)
        else:
            dt = datetime.fromisoformat(str(val).replace("Z", "+00:00"))
        return dt.isoformat()
    except (ValueError, OSError):
        return None


def build_normalized_article(
    title: str,
    url: str,
    excerpt: str | None,
    author: str | None,
    published_at: str | int | float | datetime | None,
    source: Source,
) -> NormalizedArticle | None:
    title = (title or "").strip()
    if not title or not url:
        return None

    iso = to_iso_string(published_at)
    if not iso:
        return None

    body_excerpt = normalize_excerpt(excerpt)
    raw_category = categorize(title, body_excerpt)

    return NormalizedArticle(
        title=title,
        body_excerpt=body_excerpt,
        source_name=source.name,
        source_url=url,
        author=author,
        published_at=iso,
        raw_category=raw_category,
        significance_base=float(source.authority),
    )
